### DISCUSSION TITLE
What's the advantages of response model validation in production at cost of performance?

### DISCUSSION THREAD

**Discussion by sjquant** [fastapi/fastapi#5393] [Questions] [answered]
Hello, FastAPI guys.

I have a question, and want to know your opinions.
FastAPI provides very easy way to validate input and output with the help of Pydantic.

I understand why runtime input validation is important and how easy FastAPI makes it.
But I still don't understand why output validation is required in production.

In my opinion, when we validate user input, then we can use static type checker like mypy or pyright, and ensure that output type is right. But, as far as I know, FastAPI forces response model validation. (I'm not sure if we can turn off validation with config or something. Can we do that?)

I know pydantic is quite fast, and will be more fast with v2.0. However, we will have **some performance overhead** with validation logic in any way. I think that in development phase, response validation might help because we can easily notice wrong response. In production phase, however, I don't think we need to validate all requests at cost of some overhead.

I think I might've missed some points, so I want to know your opinions and the author's opinion if possible(@tiangolo).
Thanks in advance.

**Comment by StephenPCG** [comment, marked as answer]
Hi @sjquant,

I'm not sure what do you mean by "turn off validation". IMO, `response_model` is a serializer more than a validator, you often have to serialize a database model into a json string, and pydantic model does this well. If you have other method to do serialization, you can of course "turn off validation", for example, set `response_model` to `dict` or even leave it unset, and serialize data by yourself. 

BTW, I think serialization can hardly be done without validation, especially when there are nested structures, datatype converting is itself a kind of validation. Use pydantic model for `response_model` brings a lot of benefits, e.g  documentation, serialization, error discovering in development phase (and sometimes in production phase when some data are user input related). 

It's true that `response_model` performance is bad, it has to first convert orm model to a pydantic model, then to a dict, and finally serialized into json. You can try the author's another great project [SQLModel](https://sqlmodel.tiangolo.com/), which is both an orm model and pydantic model, and thus make the serialization routine short.

**Reply by sjquant** [reply]
Thank you for your answer. It helped me understand more context. @StephenPCG 

I meant "turn off validation" as "skip validation". So, using pydantic in FastAPI in response is for seralizing response rather than validating response, and other benefits like documentation, ...

What do you think of this option?

When it comes to costly responses like large list of dict, just setting response_model with pydantic model for documentation, and use custom response class to serialize and skip converting-to-pydantic and validating process?

**Reply by StephenPCG** [reply]
@sjquant Of course you can do that. We're using this technic in some interfaces, but for performing further data tweaking rather than skipping validation. Here is an example (building pagination response, pseudo code, just for demonstration):

```py
class PageResponseModel(GenericModel, Generic[PageItemModel]):
    data: List[PageItemModel]
    pagination: Pagination

@router(
    "/api/foo",
    # since the function returns Response class directly, this is just for documentation
    response_model=PageResponseModel[FooModel],
)
def list_foo_api(page: int = Query(1), page_size: int = Query(10), filter_conditions: ...):
    query = session.query(Foo).filter(filter_conditions...)
    # PageResponse is a subclass of JSONResponse, so when returning PageResponse, fastapi will not do any serialization or validation
    return PageResponse(item_model=FooModel, query=query, page=page, page_size=page_size)

class PageResponse(JSONResponse):
    def __init__(self, item_model, query, page, page_size):
        # use page and page_size to build sql query based on the passed in base query, and construct a Pagination object
        items = query.limit(page_size).offset((page - 1) * page_size).all()
        total = query.order_by(None).count()
        pagination = Pagination(total=total, page=page, page_size=page_size)

        # convert data to response model and serialize, of course you can serialize data in your own way
        content = jsonable_encoder(parse_obj_as(PageResponseModel[item_model], items))
        super().__init__(content=content)
```

**Reply by sjquant** [reply]
Thank you for taking your time. It was very helpful. 👍
